python/ray/experimental/sgd/pytorch/pytorch_trainer.py

```diff
@@ -74,67 +76,82 @@ def __init__(self,
                  "https://github.com/pytorch/examples/issues/467."))
 
         self.model_creator = model_creator
+        self.data_creator = data_creator
         self.train_function = train_function
+        self.optimizer_creator = optimizer_creator
+        self.loss_creator = loss_creator
         self.validation_function = validation_function
+        self.initialization_hook = initialization_hook
         self.config = {} if config is None else config
         self.optimizer_timer = utils.TimerStat(window_size=1)
 
         if backend == "auto":
             backend = "nccl" if use_gpu else "gloo"
 
         logger.info("Using {} as backend.".format(backend))
-
+        self.backend = backend
+        self.use_gpu = use_gpu
+        self.batch_size = batch_size
+        self.max_replicas = num_replicas
+        self.tmpdir = tempfile.mkdtemp(prefix="raysgd")
+        self._num_failures = 0
+        self._start_workers(self.max_replicas)
+
+    def _start_workers(self, num_replicas):
+        logger.info(f"start_workers: Setting {num_replicas} replicas.")
         if num_replicas == 1:
             # Generate actor class
             Runner = ray.remote(
-                num_cpus=1, num_gpus=int(use_gpu))(PyTorchRunner)
+                num_cpus=1, num_gpus=int(self.use_gpu))(PyTorchRunner)
             # Start workers
             self.workers = [
                 Runner.remote(
-                    model_creator,
-                    data_creator,
-                    optimizer_creator,
-                    loss_creator,
-                    train_function=train_function,
-                    validation_function=validation_function,
+                    self.model_creator,
+                    self.data_creator,
+                    self.optimizer_creator,
+                    self.loss_creator,
+                    train_function=self.train_function,
+                    validation_function=self.validation_function,
                     config=self.config,
-                    batch_size=batch_size)
+                    batch_size=self.batch_size)
             ]
-            if initialization_hook:
-                self.apply_all_workers(initialization_hook)
+            if self.initialization_hook:
+                self.apply_all_workers(self.initialization_hook)
             # Get setup tasks in order to throw errors on failure
             ray.get(self.workers[0].setup.remote())
         else:
             # Generate actor class
             Runner = ray.remote(
-                num_cpus=1, num_gpus=int(use_gpu))(DistributedPyTorchRunner)
+                num_cpus=1,
+                num_gpus=int(self.use_gpu))(DistributedPyTorchRunner)
+            Runner = Runner.options(is_direct_call=True, max_concurrency=2)
```

Comment:
Add a comment saying why is max_concurrency=2?

Reply:
good catch